Add query blocking proposal (#5029)
grafana/mimir: docs/proposals/query-blocking.md

---

> title: "XXXX: Query blocking"
> description: "Query blocking"
> draft: true
> ---
> 
> # XXXX: Query blocking

Reviewer: ```suggestion
title: "0001: Query blocking"
description: "Query blocking"
draft: true
---

# 0001: Query blocking
```

borrowing from what loki is doing

**Author:** @wilfriedroset

**Date:** 05/2023

> **Sponsor(s):** @username of maintainer(s) willing to shepherd this MID

Reviewer: ```suggestion
**Sponsor(s):** @dimitarvdimitrov 
```

I'm happy to help with delivering this

**Type:** Feature

**Status:** Draft

**Related issues/PRs:**

---

## Background

The `query-frontend` is a component that is responsible for query acceleration through splitting, sharding, caching, and other techniques. It is also responsible for enforcing limits on queries such as total query time range.

## Problem Statement

In certain situations, you may not be able to control the queries being sent to your Mimir installation. These queries may be intentionally or unintentionally expensive to run, and they may affect the overall stability or cost of running your service.

Example:

```
sum(increase(tfe_run_current_count{status=~"applied|planned_and_finished", instance="host1", organization_name=~".*", workspace_name=~".*"}[2674800s]) > 0)
```

> The query above can be harmful as the wildcard regex would match any label values including empty string. The result of such query is not cached by Mimir. As such the read path would be under pressure if the query is repeated by the tenant (e.g: dashboard in auto repeat).

Reviewer: > The result of such query is not cached by Mimir.

What's the reason for this? nothing strikes me as uncacheable here, maybe the large period of range selector?

Author: Our analysis was wrong about this point. Mimir was under too much pressure.
However, the need for this feature remains, I've updated the proposal to remove this misleading paragraph.

## Goals

- Prevent harmful queries to be processed by Mimir's read path.

## Non-Goals (optional)

- Rewrite queries in a non harmful way.
- Optimize queries on the fly.
- Blocked remote read.

> - Block similar queries with a single definition. Blocking `metric_name{label1=value1, label2=value2}` does not automatically block `metric_name{label2=value2, label1=value1}`.

Reviewer: I wonder if this shouldn't be the goal to some extent. I think avoiding whitespace changes at least makes sense. At the same time, I understand if as a first iteration you don't want to include this "normalization"

Author: Taken into account whitespace changes seems easy but I would rather either completely normalize the queries or not at all. I reckon it is simpler to understand from an operator point of view.

- Lint or re-format queries.

## Proposals

### Proposal 0: Do nothing

Current tenants isolation through `shuffle-sharding` are sufficient to keep the read path healthy. Mimir's operators can use limits such as `max_fetched_chunks_per_query`, `max_queriers_per_tenant` or `store_gateway_tenant_shard_size` to limits tenants processing capabilities.

**Pros:**

- Nothing to be done

**Cons:**

- Harmful queries are still processed by Mimir but in a limited part of the cluster

### Proposal 1: Query blocking

> Based on the configuration, Mimir's [limitsMiddleware](https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/limits.go)) runned by the `query-frontend` would be responsible for blocking queries. Operators would be able to block queries either for all tenants or on a per-tenant basis via the _runtime configuration_.

Reviewer: ```suggestion
Based on the configuration, Mimir's [limitsMiddleware](https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/limits.go) ran by the `query-frontend` would be responsible for blocking queries. Operators would be able to block queries either for all tenants or on a per-tenant basis via the _runtime configuration_.
```

```yaml
overrides:
  "tenant-id":
    blocked_queries:
      # block this query exactly
      - pattern: 'sum(rate({env="prod"}[1m]))'

      # block any query matching this regex pattern
      - pattern: '="\.\*"' # all queries with a label value `.*`
        regex: true
```

> The order of patterns is preserved, so the first matching pattern will be used. If `blocked_queries` are defined in both _runtime_configuration_ and _configuration_, the resulting list should be the concatenation of both with the _runtime_configuration_ first.

Reviewer 2: I'm not sure about the "concatenation". Currently the runtime configuration, where you specify the per-tenant limits simply takes precedence over the standard one. I would start keeping this behaviour.

PromQL queries might contains special characters used in regexp (`[](){},-`) or regexp.
As such, a list of blocked queries, either regex or non-regex, is preferred over a single regex. This would avoid a eventual escaping problems.
To accommodate multi-lines promQL queries, Mimir's operators can use YAML multi-line strings.

> Blocked queries must be logged, as well as counted in the `mimir_blocked_queries` metric on a per-tenant basis.

Reviewer 2: ```suggestion
Blocked queries must be logged, as well as counted in the `cortex_query_frontend_blocked_queries_total` metric on a per-tenant basis.
```

## Goals

- Prevent harmful queries to be processed by Mimir's read path.

## Non-Goals (optional)

- Rewrite queries in a non harmful way.
- Optimize queries on the fly.
- Blocked remote read.

## Proposals

### Proposal 0: Do nothing

Current tenants isolation through `shuffle-sharding` are sufficient to keep the read path healthy. Mimir's operators can use limits such as `max_fetched_chunks_per_query`, `max_queriers_per_tenant` or `store_gateway_tenant_shard_size` to limits tenants processing capabilities.

**Pros:**

- Nothing to be done

**Cons:**

- Harmful queries are still processed by Mimir but in a limited part of the cluster

### Proposal 1: Query blocking

Based on the configuration, Mimir's [limitsMiddleware](https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/limits.go)) runned by the `query-frontend` would be responsible for blocking queries. Operators would be able to block queries either for all tenants or on a per-tenant basis via the _runtime configuration_.

```yaml
overrides:
  "tenant-id":

>     blocked_queries:

Reviewer 2: A couple of questions:
1. What if we just allow to configure 1 regexp matching the query, to start with an easier solution?
2. Do we really need `hash`? What's the actual use case for that?

Author: > What if we just allow to configure 1 regexp matching the query, to start with an easier solution?

Yes we could start with this simpler version. The risk here is to have a hard to read regex

> Do we really need hash? What's the actual use case for that?

Depending on the complexity of the query you want to block the configuration could be rather big. Having the possibility to block a query based on the `hash` makes it simple. More over, as an administrator, I don't need to pay attention to escaping anything if I just block a query based on the `hash` outputed in mimir's logs.

Reviewer: > I don't need to pay attention to escaping anything

are YAML multi-line strings an alternative?

Author: > are YAML multi-line strings an alternative?

yes this could be a solution. I will take into account your two comments and remove the `hash` and start with 1 regexp per tenant.

Reviewer: I think it's reasonable to have both regex and non-regex because of the escaping problem and how many characters in promQL are also special characters in regexps (`[](){},-`) and that promQL itself can contain regexp.

My concern with the hash is that it's difficult/error-prone to track which hash corresponds to which query from the operator point of view. It is also somewhat error-prone to construct a hash - do I need to remove spaces from the query, should I format it beforehand, etc.

Reviewer 3: Hi,

> My concern with the hash is that it's difficult/error-prone to track which hash corresponds to which query from the operator point of view. It is also somewhat error-prone to construct a hash - do I need to remove spaces from the query, should I format it beforehand, etc.

Hash of query could be computed by mimir and displayed with the query inside the log, this way their won't be any mistake possible.

> With or without the hash we have a similar issue when using regex or plain query. If an operator want to block foo{label1='value1', label2='value2'} I believe this person also expected that foo{ label1='value1', label2='value2' } and foo{label2='value2', label1='value1'} are blocked as well as they are the same query with a different formating.

I'm afraid I must disagree with you, even if those two queries should produce the same results, it doesn't mean they will be interacting the same way with mimir and it's dependencies (cache, object storage...), especially if you use regex as labels selector.

Regards,

Reviewer: > Hash of query could be computed by mimir and displayed with the query inside the log, this way their won't be any mistake possible.

It is still difficult when reviewing existing configuration. An operator can't know whether they still want to block a query hash. That forces them to track query:hash mappings in comments or some other form, which can be wrong or become outdated.

> if those two queries should produce the same results, it doesn't mean they will be interacting the same way with mimir and it's dependencies

Currently this order matters only for the ingester but IMO it's negligible. I think it's easier for an operator to think of this as the same query. Knowing whether it will be more expensive requires knowledge of the prometheus TSDB querying code and that's also subject to change.

Author: Thank you @dimitarvdimitrov for your feedbacks. I've updated the PR to remove the `hash` part and move to *non-goal* blocking similar queries.
WDYT about this update?

Reviewer: awesome, thanks for updating it!
